nnvm/python/nnvm/testing/check_computation.py

```diff
@@ -0,0 +1,451 @@
+                    max_discarded_frac=ng_max_discarded_frac,
+                    atol=ng_atol, rtol=ng_rtol)
+
+        except:
+            print("\ncheck_function failed while {}, here is the main graph".format(debug_stage))
+            print(main_graph.ir(join_node_attrs=['shape', 'dtype']))
+            if nnvm_res is not None:
+                print("Generated inputs:")
+                print(np_inputs)
+                print()
+            raise
+
+
+def check_numerical_grads(function, grad_input_vars, input_values, grad_values, function_value=None,
+                          delta=1e-3, max_error=1e+3, max_discarded_frac=0.1,
+                          atol=1e-2, rtol=1e-2):
+    """A helper function that checks that numerical gradients of a function are equal to
+    gradients computed in some different way.
+
+    We compute two approximations for each gradient using forward and backward differences.
+    Then we use the distance between these gradients as our estimate for the error and check if the
+    provided symbolic gradient lies within this distance from the central difference approximation
+    of the gradient.
+
+    Parameters
+    ----------
+    function
+        A function that takes inputs as keyword arguments (like `function(**input_values)`) and
+        returns a scalar result. Should accept and return numpy arrays.
+
+    grad_input_vars : List[str]
+        A list of variables with respect to which the gradients will be computed.
+
+    input_values : Dict[str, numpy.ndarray]
+        A dict assigning values to variables. Represents the point at which gradients should be
+        computed.
+
+    grad_values : Dict[str, numpy.ndarray]
+        Gradients computed using a different method.
+
+    function_value : float, optional
+        Should be equal to `function(**input_values)`.
+
+    delta : float
+        A small number used for numerical computation of partial derivatives.
+
+    max_error : float
+        Discard numerical partial derivatives whose estimated error is larger than this value.
+
+    max_discarded_frac : float
+        Allow discarding no more than this fraction of partial derivatives.
+
+    atol : float
+        Absolute tolerance.
+
+    rtol : float
+        Relative tolerance.
+    """
+
+    if function_value is None:
+        function_value = function(**input_values)
+
+    # a helper to modify j-th element of val by a_delta
+    def modify(val, j, a_delta):
+        val = val.copy()
+        val.reshape(-1)[j] = val.reshape(-1)[j] + a_delta
+        return val
+
+    # numerically compute a partial derivative with respect to j-th element of the var `name`
+    def derivative(x_name, j, a_delta):
+        modified_values = {n: modify(val, j, a_delta) if n == x_name else val
+                           for n, val in input_values.items()}
+        return (function(**modified_values) - function_value)/a_delta
+
+    for x_name in grad_input_vars:
+        grad = grad_values[x_name]
```

Comment:
We can replace these two lines with `for x_name, grad in grad_values.items():`.  Then `grad_input_vars` is no longer necessary.